jobsdb/prebackup/prebackup.go

```diff
@@ -0,0 +1,59 @@
+package prebackup
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/lib/pq"
+)
+
+// Handler alters a jobsdb table before backup happens
+type Handler interface {
```

Comment:
[mino] Can't this be defined in `jobsdb.go` instead and remove the coupling between the two packages? `DropSourceIds` can still return an unexported `struct` (i.e. `*dropSourceIds`). Thoughts?

Reply:
I don't understand the value of an unexported interface, please elaborate a bit on this proposal.

Reply:
It's just a minor software design (not blocking) question but I'd like to know your opinion about this.

To me it looks like the `Handler` interface is not needed in the `prebackup` package, you're just defining it upfront (kinda Java style).
The only thing using it there is `func DropSourceIds` but there is no reason to return an interface there, it could just return a struct instead.

It seems more "idiomatic" in Golang to have the package that uses the abstraction to
define it and to me that package is the `jobsdb` (i.e. in `WithPreBackupHandlers`).

Long story short, I'd move the interface in the `jobsdb` package and call it something like `PreBackupper` because that's what it does. `Handler` is a very generic name and it doesn't tell us what that abstraction actually means to the `jobsdb` package.

Once we do that we realize that the `prebackup` package doesn't even have to be called `prebackup`. What it does is that it gives you an utility function to drop jobs by `source_id`, we don't even care if those are transient either (not at that stage at least I think?).

So the `prebackup` could be `sourcedropper` and we could remove all references to
handlers and transient things from there and have a properly named interface in `jobsdb` instead.

These are my two cents, your code does what is supposed to do and there is nothing wrong with it, but as a reader of your code this is the impression I got :slightly_smiling_face:

Reply:
Yes, what you propose is valid, still the interface needs to be exported :)

Reply:
> Yes, what you propose is valid, still the interface needs to be exported :)

Not here where I added the comment, it would be in a different package where it would be more clear, both the name and its purpose.

> I don't consider it wrong to explicitly set the return type to the interface in question, instead of using an unexported struct type

That depends, we should do things when it makes sense, I don't think it does in this case.

You're returning an interface that is simply not needed here and IMO that led to:
* interface name that could be more specific like I'm suggesting instead of `Handler`
* interface in a package where it is not needed, leads to confusion
* more coupling between packages
* misleading name for the `prebackup` package

I haven't heard an argument that explains why it makes sense to return an interface here so my suggestion so far, along with the above, is: let's not return interfaces just because we are used to do so. I know that is very common the first years of Go, especially when coming from a different background so it's cool. That said, if you believe your approach is better then go ahead :+1: 

FYI: Here's some stuff worth [reading](https://blog.chewxy.com/2018/03/18/golang-interfaces/).

Reply:
Unfortunately, I cannot relate (yet) to [this](https://blog.chewxy.com/2018/03/18/golang-interfaces/) since it doesn't explain, or otherwise I fail to understand, why we shouldn't write types to fulfil interfaces. Until I do, my mind will continue working like [this](https://github.com/prometheus/client_golang/blob/main/prometheus/histogram.go#L193) 🤣